library/pk_wrap.c

```diff
@@ -920,47 +920,109 @@ static int pk_ecdsa_sig_asn1_from_psa(unsigned char *sig, size_t *sig_len,
     return 0;
 }
 
+/* This is the common helper used by ecdsa_sign_wrap() functions below (they
+ * differ in having PK_USE_PSA_EC_DATA defined or not) to sign using PSA
+ * functions. */
+static int ecdsa_sign_psa(mbedtls_svc_key_id_t key_id, psa_algorithm_t psa_sig_md,
+                          const unsigned char *hash, size_t hash_len,
+                          unsigned char *sig, size_t sig_size, size_t *sig_len)
+{
+    int ret = MBEDTLS_ERR_ERROR_CORRUPTION_DETECTED;
+    psa_status_t status;
+
+    status = psa_sign_hash(key_id, psa_sig_md, hash, hash_len,
+                           sig, sig_size, sig_len);
+    if (status != PSA_SUCCESS) {
+        return PSA_PK_ECDSA_TO_MBEDTLS_ERR(status);
+    }
+
+    ret = pk_ecdsa_sig_asn1_from_psa(sig, sig_len, sig_size);
+
+    return ret;
+}
+
+/* The reason for having this duplicated compared to ecdsa_sign_wrap() below is
+ * that:
+ * - opaque keys are available as long as USE_PSA_CRYPTO is defined and even
+ *   if !PK_USE_PSA_EC_DATA
+ * - opaque keys do not support PSA_ALG_DETERMINISTIC_ECDSA() */
```

Comment:
Following our discussion on Slack, it turned out that actually opaque keys cannot be easily changed to use deterministic ECDSA, because not only the key is owned externally, but also its properties. So if the application requires `ALG_ECDSA` and we set `ALG_ECDSA_DETERMINISTIC` the operations will be rejected.
This means that a proper wrapper for opaque keys will be added to the list proposed [here](https://github.com/Mbed-TLS/mbedtls/pull/7814#pullrequestreview-1492700429)

Reply:
Actually, the right thing to do in the opaque wrapper would be to look which of `ALG_ECDSA` or `ALG_ECDSA_DETERMINISTIC` is supported by the key and pick that one. (Currently we just assume that the key allows `ALG_ECDSA` which is an arbitrary and undocumented restriction.)

It's out of the original scope of this PR, but it seems like small addition, so perhaps that's something we can do as part of this clean-up. Otherwise I'll create a separate issue to track it.

Reply:
Oh right, I didn't thought about it, thanks! No need for another PR, let's do it in this one